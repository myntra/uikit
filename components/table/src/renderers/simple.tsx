import React, { PureComponent, ReactNode, isValidElement } from 'react'
import { TableMeta, FixedPosition, Column } from '../table-interface'
import classnames from './simple.module.scss'
import { Provider } from '../table-context'
import Measure, { MeasureData } from '@myntra/uikit-component-measure'

export interface Props extends BaseProps {
  config: TableMeta
  data: any[]
  getEnhancerState<T = any>(
    name: string
  ): { value: T; onChange(value: T): void }
}

const TR = (props: BaseProps) => <tr {...props} />
const TD = (props: BaseProps) => <td {...props} />

export default class SimpleTable extends PureComponent<
  Props,
  { offsets: Record<number, number> }
> {
  nextAnimationFrame: number
  state = {
    offsets: {},
  }

  tableContext = { TR, TD }

  defaultRowRenderer = {
    selector() {
      return true
    },
    render({ rowId, item, ...props }) {
      return <tr key={rowId} {...props} />
    },
  }

  warpIfNeeded(node: ReactNode, key: string, props: Record<string, any>) {
    if (isValidElement(node)) {
      if (node.type === 'td') {
        return React.cloneElement(node, { key })
      }
    }

    return (
      <td key={key} {...props}>
        {node}
      </td>
    )
  }

  getRowRenderer(rowId: number) {
    const row = this.props.config.rows.find((row) => row.selector(rowId))

    if (!row) return this.defaultRowRenderer

    return row
  }

  handleColumnMeasure = (column: Column, offset: MeasureData['offset']) => {
    const index = this.props.config.cells.indexOf(column)

    if (index >= 0) {
      this.setState(({ offsets }) => ({
        offsets: { ...offsets, [index]: offset.left },
      }))
    }
  }

  getEnhancerState(enhancer: string) {
    return this.props.getEnhancerState(enhancer)
  }

  render() {
    const {
      config,
      data,
      className,
      children,
      style,
      getEnhancerState: getEnhancerStateForColumn,
      ...props
    } = this.props
    const maxDepth = config.columnsByLevel.length

    return (
      <Provider value={this.tableContext}>
        <div className={classnames('simple', className)} style={style}>
          <div className={classnames('container')}>
            <table {...props} className={classnames('table')}>
              <thead>
                {config.columnsByLevel.map((columns, headLevel) => (
                  <tr key={headLevel}>
                    {columns.map((column) => (
                      <Measure
                        key={column.id}
                        onMeasure={({ offset }) =>
                          this.handleColumnMeasure(column, offset)
                        }
                      >
                        {({ ref, content }) => (
                          <th
                            ref={ref}
                            key={column.id}
                            rowSpan={maxDepth - column.level - column.depth}
                            colSpan={column.colSpan}
                            className={classnames({
                              fixed: typeof column.fixed !== 'undefined',
                              end: column.fixed === FixedPosition.END,
                            })}
                            style={{
                              // @ts-ignore
                              '--sticky-top-offset': headLevel * 35 + 'px',
                              '--sticky-left-offset':
                                typeof column.fixed !== 'undefined'
                                  ? content.offset.left + 'px'
                                  : 'unset',
                            }}
                          >
                            {column.renderHead()}
                            {column.enhancers.map(([enhancer, props]) =>
                              enhancer.renderHead(
                                {
                                  columnId: column.id,
                                  ...this.getEnhancerState(enhancer.name),
                                  getter: column.accessor,
                                },
                                props,
                                this.props.data
                              )
                            )}
                          </th>
                        )}
                      </Measure>
                    ))}
                  </tr>
                ))}
              </thead>

              <tbody>
                {data.map((item, rowId) =>
                  React.cloneElement(
                    this.getRowRenderer(rowId).render({
                      rowId,
                      item,
                      children: config.cells.map((column, columnIndex) => {
                        const cellProps = {
                          className: classnames({
                            fixed: typeof column.fixed !== 'undefined',
                            end: column.fixed === FixedPosition.END,
                          }),
                          style: {
                            '--sticky-left-offset':
                              typeof column.fixed !== 'undefined'
                                ? this.state.offsets[columnIndex] + 'px'
                                : 'unset',
                          },
                        }

                        return this.warpIfNeeded(
                          column.renderCell({
                            ...cellProps,
                            rowId,
                            columnId: column.id,
                            item,
                            data: item,
                            value: column.accessor(item, rowId),
                          }),
                          column.id,
                          cellProps
                        )
                      }),
                    }) as any,
                    { key: rowId }
                  )
                )}
              </tbody>
            </table>
          </div>
        </div>
      </Provider>
    )
  }
}
