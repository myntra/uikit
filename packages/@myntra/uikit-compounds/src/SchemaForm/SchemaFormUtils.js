import { compile } from './JSONSchema'

export function assert(condition, message) {
  if (!condition) throw new Error(message)
}

export function generateUISchema(schema, componentResolver) {
  assert(schema.type === 'object', 'Invalid schema, expected "object" as root type')

  return generateField({ name: '$root', path: '#', field: schema, schema, componentResolver })
}

function generateField({ name, path, field, schema, componentResolver, props }) {
  field.type && assert(typeof field.type === 'string', 'Invalid field type, expected "type" key to be string')
  if (field.type === 'object') {
    assert('properties' in field, 'Invalid "object" type field, expected "properties" key')

    const keys = Object.keys(field.properties)
    const required = new Set(field.required)
    const children = keys.map(name =>
      generateField({
        name,
        field: field.properties[name],
        schema,
        componentResolver,
        path: `${path}/properties/${name}`,
        props: {
          required: required.has(name)
        }
      })
    )

    const props = { ...field.ui, name }

    Object.keys(KEYS_TO_BE_COPIED).forEach(key => {
      const target = KEYS_TO_BE_COPIED[key] || key
      if (field[key]) props[target] = field[key]
    })
    const dependencies = prepareDependencies({ name, path, field, schema })

    return {
      name,
      path,
      component: componentResolver('SchemaForm.Object'),
      children,
      props,
      dependencies
    }
  } else if (field.type === 'array') {
    assert('items' in field, 'Invalid "array" type field, expected "items" key')

    const children = [generateField({ name, field: field.items, schema, componentResolver, path: `${path}/items` })]

    const props = { ...field.ui, name }

    Object.keys(KEYS_TO_BE_COPIED).forEach(key => {
      const target = KEYS_TO_BE_COPIED[key] || key
      if (field[key]) props[target] = field[key]
    })

    return { name, path, component: componentResolver('SchemaForm.Array'), children, props }
  } else {
    const [component, localProps] = generatePrimitive(name, path, field, componentResolver)

    return { name, path, component, props: { ...props, ...localProps } }
  }
}

const KEYS_TO_BE_COPIED = { title: 'label', description: '', pattern: '', readOnly: '', multipleOf: 'step' }

function generatePrimitive(
  name,
  path,
  { type: type$, enum: enum$, format: format$, ui = {}, ...field },
  componentResolver
) {
  const { component, ...props } = ui

  props.name = name

  Object.keys(KEYS_TO_BE_COPIED).forEach(key => {
    const target = KEYS_TO_BE_COPIED[key] || key
    if (field[key]) props[target] = field[key]
  })

  if (component) {
    const ref = componentResolver(component)

    assert(ref, `Unknown component <${component}>`)

    return [ref, props]
  }

  if (enum$) {
    if (enum$.length < 5) {
      return [componentResolver('Form.RadioGroup'), props]
    } else {
      return [componentResolver('Form.Select'), props]
    }
  }

  switch (type$) {
    case 'number':
      return [componentResolver('Form.Number'), props]
    case 'integer':
      return [componentResolver('Form.Number'), { step: 1, ...props }]
    case 'boolean':
      return [componentResolver('Form.Switch'), props]
    case 'string':
      if (!format$) return [componentResolver('Form.Text'), props]
      if (format$.startsWith('~')) return [componentResolver('Form.Select'), props]
      return [componentResolver('Form.Text'), { ...props, type: format$ }]
  }

  return []
}

function prepareDependencies({ field }) {
  if (!field.dependencies) return

  return compile({ dependencies: field.dependencies })
}
