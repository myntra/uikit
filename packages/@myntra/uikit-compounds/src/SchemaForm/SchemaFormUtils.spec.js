import { generateUISchema } from './SchemaFormUtils'

import simple from './fixtures/simple.schema.json'

function namedFn(name) {
  return () => {}
}

describe('generateUISchema', () => {
  it('should generate UI schema', () => {
    const ui = generateUISchema(simple, namedFn)

    expect(ui).toMatchSnapshot()
  })
})
