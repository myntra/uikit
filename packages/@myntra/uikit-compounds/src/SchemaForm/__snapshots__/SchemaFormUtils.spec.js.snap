// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateUISchema should generate UI schema 1`] = `
Object {
  "children": Array [
    Object {
      "component": [Function],
      "name": "name",
      "path": "#/properties/name",
      "props": Object {
        "description": "Your full name",
        "label": "Name",
        "name": "name",
        "required": true,
      },
    },
    Object {
      "component": [Function],
      "name": "email",
      "path": "#/properties/email",
      "props": Object {
        "description": "Your personal email address",
        "label": "Email Address",
        "name": "email",
        "required": true,
        "type": "email",
      },
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "component": [Function],
              "name": "street",
              "path": "#/properties/addresses/items/properties/street",
              "props": Object {
                "label": "Street",
                "name": "street",
                "required": false,
              },
            },
            Object {
              "component": [Function],
              "name": "phone",
              "path": "#/properties/addresses/items/properties/phone",
              "props": Object {
                "label": "Contact Number",
                "name": "phone",
                "required": false,
                "type": "tel",
              },
            },
            Object {
              "component": [Function],
              "name": "pin",
              "path": "#/properties/addresses/items/properties/pin",
              "props": Object {
                "label": "PIN Code",
                "name": "pin",
                "pattern": "^\\\\d{6}$",
                "required": false,
              },
            },
          ],
          "component": [Function],
          "dependencies": undefined,
          "name": "addresses",
          "path": "#/properties/addresses/items",
          "props": Object {
            "description": "Your contact address",
            "label": "Contact Address",
            "name": "addresses",
          },
        },
      ],
      "component": [Function],
      "name": "addresses",
      "path": "#/properties/addresses",
      "props": Object {
        "description": "Your all contact addresses",
        "label": "Contact Addresses",
        "name": "addresses",
      },
    },
  ],
  "component": [Function],
  "dependencies": undefined,
  "name": "$root",
  "path": "#",
  "props": Object {
    "label": "Simple Form",
    "name": "$root",
  },
}
`;
